readme.md
# node-enliven-minetest

Manage minetestserver using a convenient web interface.

Authentication for players via the mt auth db is planned
(see status at issue #2)!

Before using this webapp, make sure you have installed minetest to
$HOME/minetest with the RUN_IN_PLACE option (this option is true by
default in Final Minetest builds). In the future, this webapp should
ask you which minetest to use: track the progress of this feature at
issue #5.

The webapp must run as same user as minetestserver, and neither should
run as root.


## Install
* Using Terminal, cd to your EnlivenMinetest/webapp diretory, then:
```bash
npm install
```


## Features
- Upload skins.
- See Issue #2 for status of planned core feature set before use!


## Usage
* Start the server:
  `node server.js`\
  The server will create the public/skins directory automatically.\
* To force updating skins during startup, delete the public/skins
  directory if already exists and is outdated.
* In browser, go to http://localhost:64638
* For security, no overwriting of existing skins is allowed until
  authentication is implemented: See Issue #2 for status of the planned
  core feature set.


## Developer Notes

This project uses passport
(see <https://code.tutsplus.com/tutorials/authenticating-nodejs-applications-with-passport--cms-21619>

### Development Log
(for changelog, see CHANGELOG.md)
```bash
#!/bin/sh
sudo apt update
sudo apt install nodejs npm mongodb
# NOTE: mongo daemon is called mongod
target_dir=$HOME/enlivenode
if [ ! -d "$target_dir" ]; then
  print "ERROR: Nothing done since missing $target_dir"
  exit 1
fi
cd "$target_dir"
npm init
#except changed jade to pug
npm install express static-favicon morgan cookie-parser body-parser debug pug passport passport-local mongoose multer mv
#NOTE: multiparty has streaming like busboy, but is non-trivial to implement
```

### Old (Unused)

```
#!/bin/sh
wget https://raw.githubusercontent.com/ericf/express-handlebars/master/examples/basic/server.js
wget https://raw.githubusercontent.com/ericf/express-handlebars/master/examples/basic/package.json
mkdir views
cd views
wget https://raw.githubusercontent.com/ericf/express-handlebars/master/examples/basic/views/home.handlebars
mkdir layouts
cd layouts
wget https://raw.githubusercontent.com/ericf/express-handlebars/master/examples/basic/views/layouts/main.handlebars

if [ -d ../layouts ]; then
  cd ..
fi
if [ -d ../views ]; then
  cd ..
fi
```


changelog.md
# Changelog

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).


## [git] - 2020-04-08
### Changed
- Upgrade all packages as per
  <https://medium.com/@jh3y/how-to-update-all-npm-packages-in-your-project-at-once-17a8981860ea>:
  - add wipe-dependencies.js (trivial script; added later since forgot,
    then re-ran commands below)
  - `node wipe-dependencies.js && rm -rf node_modules && npm update --save-dev && npm update --save`


## [git] - 2020-03-15
### Changed
- Split the app into a separate project (it was formerly in
  the webapp directory of EnlivenMinetest).


## [git] - 2019-03-19
### Added
- Upload skins (the server operates based on ENLIVEN if present,
  otherwise based on Bucket_Game).
- Choose existing skins, those being skins included in coderskins
  (`skin_*.png` where `*` is not a username).
- `npm install multer`: mv is no longer necessary--I had switched from
   fs.rename to mv due to rename not working across filesystems (tmp is
   commonly on a different volume), but then switched from formidable
   to multer since multer has built-in functionality to cancel upload
   if too large, but formidable crashes (see
   <https://stackoverflow.com/a/27067596/4541104>).


js/main.js
/**
 * This is part of node-enliven-minetest, a webapp for managing
 * Minetest.
 * @module node-enliven-minetest/main
 * @license
 * Copyright (C) 2020 Jake Gustafson
 * This code and any other files where not specified otherwise are
 * subject to the license (See license.txt).
*/

var tzElement = document.getElementById("tzArea");
var tz_offset = tzElement.innerHTML;
var dateElement = document.getElementById("dateArea");
var date_s = dateElement.innerHTML;
var start_hour = 0;
var start_m = start_hour*60;
var start_s = start_m*60;
var endbefore_hour = 24;
var hour_count = endbefore_hour - start_hour;
var hour_width = 60;  // 60 pixels per hour
var pixels_per_m = hour_width/60.0;
var pixels_per_s = pixels_per_m/60.0;
var player_name_width = 100;
var text_size = 16.0;
var row_y_offset = text_size * (22.0/16.0); // if 16px font is drawn at (x,16), line at (x,21) is one pixel away from the bottom of the descenders
var border_y = 0.5; //canvas uses 0.5 as middle of pixel
var border_start_x = 0.5+player_name_width;
var border_x = border_start_x;
var text_y = text_size;
var text_start_x = 3.5;

var players = [];
var stats_html = "";
var time_area_width = hour_width*hour_count;
var canvas_w = player_name_width + time_area_width;
stats_html += '<div>' + hour_count + "hrs * " + hour_width + "px = " + time_area_width + '</div>'
stats_html += '<div>' + time_area_width + "px + " + player_name_width + "px player name = " + canvas_w + 'px canvas width</div>'
var outputElement = document.getElementById("outputArea");
var statusElement = document.getElementById("statusArea");
statusElement.innerHTML="...";
function draw_players() {
	var canvas = document.getElementById("myCanvas");
	canvas.width = canvas_w;
	var ctx = canvas.getContext("2d");
	ctx.font = text_size+"px Arial";
	var arrayLength = players.length;
	var text_x;
	var index = -1;
	for (var i = 0; index < arrayLength; i++) {
		ctx.font = text_size+"px Arial";
		text_x = text_start_x;
		ctx.fillStyle = "#dce5ea";
		ctx.fillRect(0, border_y+1.0, player_name_width, row_y_offset-1.0);
		//ctx.stroke();
		ctx.fillStyle = "#000000";
		ctx.moveTo(0, border_y);
		ctx.strokeStyle = "#000000";
		ctx.lineTo(canvas_w, border_y);

		//ctx.stroke();
		ctx.fillStyle = "#000000";
		if (index==-1) {
			ctx.font = "bold "+text_size+"px Arial";
			ctx.fillText("Name", text_x, text_y);
			ctx.font = text_size+"px Arial";
			text_x = player_name_width+0.5;

			//ctx.fillStyle = "#000000";
			ctx.strokeStyle = "#008000";
			for (var hour=0; hour<endbefore_hour; hour++) {
				ctx.fillStyle = "#B0B0B0";
				ctx.moveTo(text_x, border_y);
				ctx.lineTo(text_x, border_y+row_y_offset-0.5);
				//ctx.stroke();
				ctx.fillText(hour+":00", text_x+3, text_y);
				text_x += hour_width;
			}
		}
		else {
			ctx.fillText(players[index].display_name, text_x, text_y);
			var border_x = border_start_x;
			var playsLength = 0;
			var plays = null;
			if (date_s!=null) {
				if (players[index].plays != undefined) {
					if (players[index].plays.hasOwnProperty(date_s)) {
						plays = players[index].plays[date_s];
						playsLength = plays.length;
					}
				}
			}
			for (var s_i = 0; s_i < playsLength; s_i++) {
				if (!plays[s_i].hasOwnProperty("login_time")) {
					plays[s_i].login_time = "00:00:00";
				}
				var login_places = plays[s_i].login_time.split(':');
				if (login_places.length>=3) {
					login_places[0] -= Math.floor(tz_offset/60);
				}
				if (!plays[s_i].hasOwnProperty("logout_time")) {
					plays[s_i].logout_time = "23:59:59";
				}
				var logout_places = plays[s_i].logout_time.split(':');
				if (logout_places.length>=3) {
					logout_places[0] -= Math.floor(tz_offset/60);
				}
				var login_second = (+login_places[0]) * 60 * 60 + (+login_places[1]) * 60 + (+login_places[2]);
				var login_m = login_second / 60.0;
				var logout_second = (+logout_places[0]) * 60 * 60 + (+logout_places[1]) * 60 + (+logout_places[2]);
				var logout_m = logout_second / 60.0;
				var left_x = border_x + (login_m-start_m)*pixels_per_m;
				var right_x = border_x + (logout_m-start_m)*pixels_per_m;
				if (right_x<left_x) {
					right_x = canvas_w; //not logged out yet, so end of bar is at end
				}
				ctx.fillStyle = "#008000";
				ctx.fillRect(left_x, border_y+(row_y_offset/3.0), right_x-left_x, row_y_offset/3.0);
				ctx.font = (text_size*.66)+"px Arial";
				ctx.fillStyle = "#000000";
				ctx.fillText(login_places[0]+":"+login_places[1], left_x, text_y-(row_y_offset/3.0));
				ctx.fillStyle = "#808080";
				var logout_s = logout_places[0]+":"+logout_places[1];
				ctx.fillText(logout_s, right_x-ctx.measureText(logout_s).width, text_y+(row_y_offset/3.0));
			}
		}
		ctx.stroke();
		text_y += row_y_offset;
		border_y += row_y_offset;
		index++;
	}
}
outputElement.innerHTML = stats_html;
//NOTE: window.location.href is the entire address including query params!
var request = new XMLHttpRequest();
var getUrl = window.location;
var request_href = getUrl.protocol + "//" + getUrl.host + "/"+'get-players';
statusElement.innerHTML = request_href+"...";
request.open('GET', request_href, true);

request.onload = function() {
  if (this.status >= 200 && this.status < 400) {
	// Success!
	statusElement.innerHTML = "loaded player(s)";
	if (!statusElement.classList.contains('alert') ) {
		statusElement.classList.add('alert');
	}
	try {
		players = JSON.parse(this.response);
		statusElement.classList.add('alert-success');
		statusElement.innerHTML = "loaded " + players.length + ' player(s) <span style="color:white">from "'+this.responseURL+'"</span>';
		draw_players();
	}
	catch(e) {
		statusElement.classList.add('alert-warning');
        statusElement.innerHTML = e + " for URL '"+request_href+"'";
        //alert(this.response);
    }
  } else {
	// We reached our target server, but it returned an error
	statusElement.innerHTML = "Error " + this.status + " accessing " + request_href;
  }
};

request.onerror = function() {
  // There was a connection error of some sort
};

request.send();


masterserver.js
/**
 * This is part of node-enliven-minetest, a webapp for managing
 * Minetest.
 * @module node-enliven-minetest
 * @license
 * Copyright (C) 2020 Jake Gustafson
 * This code and any other files where not specified otherwise are
 * subject to the license (See license.txt).
*/

var express = require('express');
var app = express();
var mt = require('./minetestinfo.js');
app.get('/get-players', function (req, res) {
    res.setHeader('Content-Type', 'application/json');
    res.send(JSON.stringify(mt.players(true)));
});

var previousAnnounceStr = "none";

app.get('/last-announce', function (req, res) {
    res.setHeader('Content-Type', 'text/plain');
    res.send(previousAnnounceStr);
});

app.get('/announce', function (req, res) {
    previousAnnounceStr = JSON.stringify(req.body);
    console.log("announce got:" + previousAnnounceStr);
    res.setHeader('Content-Type', 'text/plain');
    res.send();
});


var server = app.listen(3000, function () {
    //console.log('express-handlebars example server listening on: 3000');
    var host = server.address().address;
    var port = server.address().port;
    console.log("listserver address:");
    console.log(JSON.stringify(server.address()));
    console.log("(experimental WIP) Minetest master server is listening at http://%s:%s", host, port);
});


server.js
/**
 * This is part of node-enliven-minetest, a webapp for managing
 * Minetest.
 * @module node-enliven-minetest
 * @license
 * Copyright (C) 2020 Jake Gustafson
 * This code and any other files where not specified otherwise are
 * subject to the license (See license.txt).
*/
'use strict';
// Howto: see README.md

//function getUserHome() {
    //return process.env[(process.platform == 'win32') ? 'USERPROFILE' : 'HOME'];
//}

var tzOffsetMinutes = 240; //subtract this from server time to get local time; 4hrs is 240; 5hrs is 300
// TODO: handle tzOffsetMinutes not divisible by 60
// var selectedDateStr = null;
// selectedDateStr = "2018-05-08";

const express = require('express'),
    multer = require('multer'),
// var exphbs  = require("express-handlebars");
// exphbs  = require('../../'); // "express-handlebars"
    cookieParser = require('cookie-parser'),
    bodyParser = require('body-parser'),
    //session = require('express-session'),
    fs = require('fs'),
    readlines = require('n-readlines');
const os = require('os');
//var formidable = require('formidable');
const querystring = require("querystring");  // built-in
// TODO: var config = require(storagePath + '/config.js') // config file contains all tokens and other private info
// var fun = require('./functions.js'); // functions file contains our non-app-specific functions including those for our Passport and database work
var mt = require('./minetestinfo.js'); // functions file contains our non-app-specific functions including those for our Passport and database work

// var util = require('util')

var app = express();
app.set('view engine', 'ejs');
// see https://medium.com/@TheJesseLewis/how-to-make-a-basic-html-form-file-upload-using-multer-in-an-express-node-js-app-16dac2476610
const port = process.env.PORT || 64638;
app.use(bodyParser.urlencoded({extended:false}));  // handle body requests
app.use(bodyParser.json());  // make JSON work
app.use('/public', express.static(__dirname + '/public'));

//app.engine('handlebars', exphbs({defaultLayout: 'main'}));
//app.set('view engine', 'handlebars');
var players = [];
var playerIndices = {};
var activityDates = [];
//see https://developer.mozilla.org/en-US/docs/Learn/Server-side/Express_Nodejs/Introduction
var previousLine = null;

//#region derived from mtsenliven.py
var msgPrefixFlags = ["WARNING[Server]: ", "ACTION[Server]: "]
var msgPrefixLists = {}  // where flag is key
var mfLen = msgPrefixFlags.length;
for (var mfIndex=0; mfIndex<mfLen; mfIndex++) {
    msgPrefixLists[msgPrefixFlags[mfIndex]] = [];
}

var nonUniqueWraps = [];
nonUniqueWraps.push({
    "opener":"active block modifiers took ",
    "closer":"ms (longer than 200ms)"
});
var uniqueFlags = [
    "leaves game",
    "joins game"
];
//#endregion derived from mtsenliven.py
const skinStorage = multer.diskStorage({
    // see https://medium.com/@TheJesseLewis/how-to-make-a-basic-html-form-file-upload-using-multer-in-an-express-node-js-app-16dac2476610
    destination: function(req, file, next) {
        next(null, mt.skinsPath());  // or something like './public/photo-storage'
    },
    limits: {
        fileSize: 1*1024*1024  // in bytes
    },

    // Change filename
    filename: function(req, file, next) {
        const ext = "png";
        // var errMsg = null;
        console.log("* Checking name...");
        if (!req.body.userName) {
            return next(new Error("userName is missing"));
        }
        else if (req.body.userName.length < 1) {
            return next(new Error("userName is blank."));
        }
        if (file.size < 1)  {
            return next(new Error("image not selected"));
        }

        //if (errMsg === null) {
        var directName = "player_" + req.body.userName + '.' + ext;
        console.log("* Renaming '" + file + "' to " + directName);
        next(null, directName);
        //}
        //else {
            //console.log(errMsg);
            //next(new Error(errMsg));
        //}
        // const ext = file.mimetype.split('/')[1];
        // next(null, file.fieldname + '-' + Date.now() + '.'+ext);
    }
});
// see https://medium.com/@bmshamsnahid/nodejs-file-upload-using-multer-3a904516f6d2
const skinUpload = multer({
    storage: skinStorage,

    fileFilter: function(req, file, next) {
        const ext = "png";
        console.log("filtering...");
        var errMsg = null;
        // NOTE: return with error aborts the upload.
        if (!file) {
            errMsg = "You did not select a file.";
            req.fileValidationError = errMsg;
            return next(new Error(errMsg));  // next(null, false, new Error(errMsg))
        }
        else if (file.size < 1) {
            errMsg = "Empty file.";
            req.fileValidationError = errMsg;
            return next(new Error(errMsg));
        }

        if (!file.mimetype.startsWith('image/')) {
            errMsg = "* ERROR: file type " + file.mimetype
                        + " is not supported";
            req.fileValidationError = errMsg;
            return next(new Error(errMsg));
        }
        if (errMsg === null) {
            var directName = "player_" + req.body.userName + '.' + ext;
            console.log("* " + file.mimetype + " '" + file + "' uploaded...");
            // set player skin name to new file name:
            mt.setSkin(req.body.userName, directName);
            // TODO: allow setSkin output to res
            next(null, true);
        } else {
            console.log(errMsg);
            return next(new Error(errMsg));
        }
    }
});

function processLogLine(line, lineNumber) {
    //selectedDateStr
    //TODO: use storeUniqueLogData instead of this function
    var playerName = null;
    var verb = "";
    var timeStr = "";
    var dateStr = "";
    var playerIP = null;
    const timeStartInt = 11;
    var ufLen = uniqueFlags.length;
    mfLen = msgPrefixFlags.length;
    var verbIndex = -1;
    var verbNumber = -1;
    var msgPrefixIndex = -1;
    var msgPrefixNumber = -1;
    var msgprefix = null;
    var indexMsg = "";
    for (var mfIndex=0; mfIndex<mfLen; mfIndex++) {
        msgPrefixIndex = line.indexOf(msgPrefixFlags[mfIndex]);
        if (msgPrefixIndex > -1) {
            msgPrefixNumber = mfIndex;
            msgprefix = msgPrefixFlags[mfIndex];
            break;
        }
    }
    var skipDateEnable = false;
    for (var ufIndex=0; ufIndex<ufLen; ufIndex++) {
        verbIndex = line.indexOf(uniqueFlags[ufIndex]);
        if (verbIndex > -1) {
            verbNumber = ufIndex;
            verb = uniqueFlags[ufIndex];
            dateStr = line.substring(0,10).trim();
            //if (selectedDateStr==null || selectedDateStr==dateStr) {
                //console.log("(verbose message in processLogLine) using '" + dateStr + "' since selected '"+selectedDateStr+"'");
                timeStr = line.substring(timeStartInt, timeStartInt+8);
                //console.log("using time "+timeStr);
                if (msgprefix!=null) {
                    playerName = line.substring(msgPrefixIndex+msgprefix.length, verbIndex).trim();
                    var ipFlag = " [";
                    var ipIndex = playerName.indexOf(ipFlag);
                    if (ipIndex > -1) {
                        playerIP = playerName.substring(ipIndex+ipFlag.length, playerName.length-1);
                        playerName = playerName.substring(0, ipIndex);
                    }
                }
                else {
                    playerName = "&lt;missing msgprefix&rt;";
                }
            //}
            //else {
            //  skipDateEnable = true;
                //console.log("WARNING in processLogLine: skipping '" + dateStr + "' since not '"+selectedDateStr+"'");
            //}
            break;
        }
    }
    var index = -1;  // player index
    if (playerName != null) {
        if (playerName.length > 0) {
            if (playerIndices.hasOwnProperty(playerName)) {
                index = playerIndices[playerName];
                indexMsg = "cached ";
            }
            else {
                index = players.length;
                //players.push({});
                players[index] = {};
                players[index].displayName = playerName;
                playerIndices[playerName] = index;
                //console.log("created new index "+index);
            }
        }
        else {
            console.log("WARNING in processLogLine: zero-length player name");
        }
    }
    if (index<0 && (verb=="leaves game"||verb=="joins game")) {
        console.log("(ERROR in processLogLine) " + indexMsg +
                    "index was '"+index+"' but date was present '" +
                    dateStr + "' for '"+line+"' (no player found, but" +
                    "verb is a player verb).");
    }
    var playDateEnable = false;
    if (verb == "leaves game") {
        if (index > -1) {
            var playIndex = -1;
            if (!players[index].hasOwnProperty("plays")) {
                players[index].plays = {};
            }
            if (!players[index].plays.hasOwnProperty(dateStr)) {
                //leave login time blank--player must have logged in before the available part of the log began
                players[index].plays[dateStr] = [];
                players[index].plays[dateStr].push({});
                playIndex = 0;
            }
            else {
                if (players[index].plays[dateStr].length==0) players[index].plays[dateStr].push({});
                playIndex = players[index].plays[dateStr].length - 1;
                if (players[index].plays[dateStr][playIndex].hasOwnProperty("logoutTime")) {
                    //If last entry is incomplete, start a new one:
                    players[index].plays[dateStr].push({});
                    playIndex++;
                }
            }
            players[index].plays[dateStr][playIndex].logoutTime = timeStr;
            playDateEnable = true;
        }
    }
    else if (verb == "joins game") {
        if (index > -1) {
            if (playerIP!=null) {
                players[index].playerIP = playerIP;
                var playIndex = -1;
                if (!players[index].hasOwnProperty("plays")) {
                    players[index].plays = {};
                }
                if (!players[index].plays.hasOwnProperty(dateStr)) {
                    players[index].plays[dateStr] = [];
                    playIndex = 0;
                }
                else playIndex = players[index].plays[dateStr].length;
                players[index].plays[dateStr].push({});
                //console.log(verb+" on "+dateStr+" (length "+players[index].plays[dateStr].length+") play "+playIndex+"+1 for player ["+index+"] "+playerName+"...");
                players[index].plays[dateStr][playIndex].loginTime = timeStr;
                playDateEnable = true;
            }
            // else redundant (server writes " joins game " again
            // and shows list of players instead of ip).
            //TODO: else analyze list of players to confirm in case player logged in all day
        }
    }
    if (playDateEnable) {
        if (dateStr.length>0) {
            if (activityDates.indexOf(dateStr) < 0) {
                activityDates.push(dateStr);
            }
        }
    }
}

function storeUniqueLogData(output, lineNumber, errFlag=false) {
    var ret = "";
    var outputTrim = output.trim();
    var uPrefix = "active block modifiers took ";
    var uSuffix = "ms (longer than 200ms)";
    // (outBytes is bytes)
    var showEnable = true;
    var foundFlag = null;
    var fIndex = null;
    var alwaysShowEnable = false;
    var msgMsg = "previous message";
    var ufLen = uniqueFlags.length;
    for (var ufIndex=0; ufIndex<ufLen; ufIndex++) {
        if (output.includes(uniqueFlags[ufIndex])) {
            alwaysShowEnable = true;
        }
    }
    if (!alwaysShowEnable) {
        var mfLen = msgPrefixFlags.length;
        for (var mfIndex=0; mfIndex<mfLen; mfIndex++) {
            // such as '2018-02-06 21:08:06: WARNING[Server]: Deprecated call to get_look_yaw, use get_look_horizontal instead'
            // or 2018-02-06 21:08:05: ACTION[Server]: [playereffects] Wrote playereffects data into /home/owner/.minetest/worlds/FCAGameAWorld/playereffects.mt.
            fIndex = output.find(msgPrefixFlags[mfIndex]);
            if (fIndex >= 0) {
                foundFlag = msgPrefixFlags[mfIndex];
                break;
            }
        }
        if (foundFlag!=null) {
            var subMsg = output.substring(fIndex+flag.length).trim();
            var nUWLen = nonUniqueWraps.length;
            for (var nUWIndex=0; nUWIndex<nUWLen; nUWIndex++) {
            //for (wrap in nonUniqueWraps) {
                var wrap = nonUniqueWraps[nUWIndex];
                if (subMsg.includes(wrap["opener"]) && subMsg.includes(wrap["closer"])) {
                    subMsg = wrap["opener"] + "..." + wrap["closer"];
                    msgMsg = "similar messages";
                    break;
                }
            }
            if (msgPrefixLists[foundFlag].indexOf(subMsg) > -1) {
                showEnable = false;
            }
            else {
                msgPrefixLists[foundFlag].push(subMsg);
            }
        }
    }
    if (showEnable) {
        ret = outputTrim;
        if (foundFlag != null) {
            ret += "\n  [ EnlivenMinetest ] " + msgMsg + " will be suppressed";
        }
    }
    return ret;
}

function readLog() {
    if (players==null) players = [];
    if (playerIndices==null) playerIndices = {};
    // os.homedir() + "/.minetest/debug_archived/2018/05/08.txt",
    // var logPaths = [os.homedir() + "/.minetest/debug.txt"];
    var logPaths = [os.homedir() + "/minetest/bin/debug.txt"];
    var lpLen = logPaths.length;
    for (var lpIndex=0; lpIndex<lpLen; lpIndex++) {
        var thisLogPath = logPaths[lpIndex];
        console.log("EnlivenMinetest webapp reading '" + thisLogPath + "'...");
        var lineNumber = 1;
        if (fs.existsSync(thisLogPath)) {
            //uses n-readlines package: see https://stackoverflow.com/questions/34223065/read-lines-synchronously-from-file-in-node-js
            var readLines = new readlines(thisLogPath);
            var nextLine = true;
            while (nextLine) {
                nextLine = readLines.next();
                if (nextLine!=false) {
                    processLogLine(nextLine.toString('ascii'), lineNumber);
                    lineNumber++;
                }
            }
        }
        else {
            console.log("WARNING: file not found: '" + thisLogPath + "' (listing actual archived log folders is not yet implemented, so this is a hard-coded demo folder only on poikilos' server)");
        }
    }
}

app.get('/modding', function(req, res, next) {
    res.render('pages/modding', {
        msg: "",
    });
});
app.get('/getting-started', function(req, res, next) {
    res.render('pages/getting-started', {
        msg: "",
    });
});
app.get('/skin-upload-form', function(req, res, next) {
    //var ending = "";
    //ending += '<a href="/">Back to Main Site</a><br/>' + "\n";
    ////ending += '<a href="/skin-upload-form">Back to Upload</a><br/>' + "\n";
    //ending += '</body></html>';
    //res.write('<html><body style="font-family:calibri,sans">'+"\n");
    //res.write('<form action="/upload-skin" method="post" enctype="multipart/form-data">'+"\n");
    //res.write('User Name (case-sensitive): <input type="text" name="userName" id="userName">'+"\n");
    //res.write('Select a png image to upload:'+"\n");
    //res.write('<input type="file" name="userFile" id="userFile">'+"\n");
    //res.write('<input type="submit" value="Upload Image" name="submit">'+"\n");
    //res.write('</form>'+"\n");
    //res.end(ending);
    var msg = "";
    res.render('pages/skin-upload-form', {
        msg: msg
    });
});
app.get('/skin-selection-form', function(req, res, next) {
    var msg = "";
    res.render('pages/skin-selection-form', {
        msg: msg,
        skinFileNames: mt.selectableSkinFileNames()
    });
});


// see "new way" of handling multer errors: https://github.com/expressjs/multer#error-handling
var singleSkinUpload = skinUpload.single('userFile');
app.post('/upload-skin', function(req, res) {
    singleSkinUpload(req, res, function(err) {
        if (err instanceof multer.MulterError) {
            // A Multer error occurred when uploading.
            res.render('multer error', { error: err });
            //res.render('pages/result', {
                //msg: "An error occurred in processing the form: " + err,
            //});
        } else if (err) {
            // An unknown error occurred when uploading.
            res.render('unknown error', { error: err });
            //res.render('pages/result', {
                //msg: "An error occurred in processing: " + err,
            //});
        }
        else {
            // var ending = "";
            var msg = "";
            // ending += '<a href="/">Back to Main Site</a><br/>' + "\n";
            // ending += '<a href="/skin-upload-form">Back to Upload</a><br/>' + "\n";
            // ending += '</body></html>';
            // res.write('<html>');
            // res.write('<body style="font-family:calibri,sans">');
            if (!req.fileValidationError) {
                // res.write('<p>Complete!</p>');
                msg = "Complete!";
            }
            else {
                msg = req.fileValidationError;
                // res.write('<p>' + req.fileValidationError + '</p>');
            }
            //res.end(ending);
            res.render('pages/result', {
                msg: msg
            });
        }
    })
});

app.get('/select-skin', function(req, res, next) {
    mt.setSkin(req.query.userName, req.query.skinFileName);
    res.render('pages/result', {
        msg: "Complete."
    });
});


app.get('/', function(req, res, next) {
    //var ret = "";

    //res.write('<html>');
    //res.write('<body style="font-family:calibri,sans">');
    // Whenever server starts the following is logged
    // (see also etc/example-input.txt):
    //
    //-------------
    //  Separator
    //-------------
    //
    var selectedDateStr = null;
    var msg = "";
    if (req.query.date) selectedDateStr = req.query.date
    if (req.query.msg != undefined) {
        //res.write("<br/>");
        //res.write("<b>" + querystring.parse(req.query.msg) + "</b><br>\n");
        // line above causes:
        //TypeError: Cannot convert object to primitive value
        //at /home/owner/git/EnlivenMinetest/webapp/server.js:390:16
        //at Layer.handle [as handle_request] (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/layer.js:95:5)
        //at next (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/route.js:137:13)
        //at Route.dispatch (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/route.js:112:3)
        //at Layer.handle [as handle_request] (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/layer.js:95:5)
        //at /home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/index.js:281:22
        //at Function.process_params (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/index.js:335:12)
        //at next (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/index.js:275:10)
        //at expressInit (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/middleware/init.js:40:5)
        //at Layer.handle [as handle_request] (/home/owner/git/EnlivenMinetest/webapp/node_modules/express/lib/router/layer.js:95:5)

        //res.write("<br/>");
    }
    //res.write('<p><a href="/skin-upload-form">Upload Skin</a></p>');
    //res.write("<h3>Server info</h3>");
    //res.write("<ul>");
    //res.write("<li>assuming minetestserver ran as: " + os.homedir() + "</li>");
    //res.write("<li>timezone (tzOffsetMinutes/60*-1): " + (Math.floor(tzOffsetMinutes/60)*-1) + '<span name="tzArea" id="tzArea"></span>' + "</li>");
    //res.write('<li>date: <span id="dateArea" name="dateArea">' + ((selectedDateStr!=null)?selectedDateStr:"not selected") + '</span>' + "</li>");
    //res.write('var activityDates = [];');
    var pdLength = 0;
    if (activityDates != null) pdLength = activityDates.length;
    var logDates = [];
    for (var pdIndex = 0; pdIndex < pdLength; pdIndex++) {
        //res.write('activityDates.push("' + activityDates[pdIndex] + '");');
        if (selectedDateStr!=activityDates[pdIndex]) {
            //res.write('<a href="?date='+activityDates[pdIndex]+'">'+activityDates[pdIndex]+'</a> ');
            logDates.push( { date:activityDates[pdIndex], active:true});
        }
        else {
            logDates.push( { date:activityDates[pdIndex], active:false});
            //res.write(activityDates[pdIndex]+' ');
        }
    }
    //if (selectedDateStr==null) {
        //res.write('<a href="?date=2018-05-08">2018-05-08</a>');
    //}
    //see ~/.minetest/debug.txt
    //and logs archived by EnlivenMinetest:
    //~/.minetest/debug_archived/2018/05/08.txt
    //res.write('</body></html>');
    //res.end('</body></html>');
    //res.render('home');

    res.render('pages/index', {
        msg: msg,
        serverUser: os.homedir(),
        tzOffset: (Math.floor(tzOffsetMinutes/60)*-1),
        selectedDate: ((selectedDateStr!=null)?selectedDateStr:"not selected"),
        logDates: logDates
    });
});



var server = app.listen(port, function () {
    // 8123 is default for Minecraft DynMap
    // 64638 spells 'minet' on a telephone keypad, but 6463, 6472 is already Discord RPC server
    //console.log('express-handlebars example server listening on: ' + port);
    var host = server.address().address;
    var port = server.address().port;
    console.log("server address:");
    console.log(JSON.stringify(server.address()));
    console.log("reading log...");
    readLog();
    console.log("EnlivenMinetest webapp is listening at http://%s:%s", host, port);
});
